### PR TITLE
performance optimizations.

### PR DESCRIPTION
**Pull request by Kvadratni**
This fiddle will show you the performance improvements
https://jsfiddle.net/by4Lh9fu/1/
The biggest changes are:
1. Uppercasing the constant right away so we don't have to do it every time.
2. merging the 2 filter functions as they don't act as streams in javascript.
3.  removing the render of words before the first letter input

### CODE REVIEW

File: wordlespace.html [hunk 1]
```diff
@@ -1,318 +1,23 @@
...
...", "YRENT", "YRIVD", "YRNEH", "YSAME", "YTOST", "YUANS", "YUCAS", "YUCCA", "YUCCH", "YUCKO", "YUCKS", "YUCKY", "YUFTS", "YUGAS", "YUKED", "YUKES", "YUKKY", "YUKOS", "YULAN", "YULES", "YUMMO", "YUMMY", "YUMPS", "YUPON", "YUPPY", "YURTA", "YURTS", "YUZUS", "ZABRA", "ZACKS", "ZAIDA", "ZAIDY", "ZAIRE", "ZAKAT", "ZAMAN", "ZAMBO", "ZAMIA", "ZANJA", "ZANTE", "ZANZA", "ZANZE", "ZAPPY", "ZARFS", "ZARIS", "ZATIS", "ZAXES", "ZAYIN", "ZAZEN", "ZEALS", "ZEBEC", "ZEBRA", "ZEBUB", "ZEBUS", "ZEDAS", "ZEINS", "ZENDO", "ZERDA", "ZERKS", "ZEROS", "ZESTS", "ZESTY", "ZETAS", "ZEXES", "ZEZES", "ZHOMO", "ZIBET", "ZIFFS", "ZIGAN", "ZILAS", "ZILCH", "ZILLA", "ZILLS", "ZIMBI", "ZIMBS", "ZINCO", "ZINCS", "ZINCY", "ZINEB", "ZINES", "ZINGS", "ZINGY", "ZINKE", "ZINKY", "ZIPPO", "ZIPPY", "ZIRAM", "ZITIS", "ZIZEL", "ZIZIT", "ZLOTE", "ZLOTY", "ZOAEA", "ZOBOS", "ZOBUS", "ZOCCO", "ZOEAE", "ZOEAL", "ZOEAS", "ZOISM", "ZOIST", "ZOMBI", "ZONAE", "ZONAL", "ZONDA", "ZONED", "ZONER", "ZONES", "ZONKS", "ZOOEA", "ZOOEY", "ZOOID", "ZOOKS", "ZOOMS", "ZOONS", "ZOOTY", "ZOPPA", "ZOPPO", "ZORIL", "ZORIS", "ZORRO", "ZOUKS", "ZOWEE", "ZOWIE", "ZULUS", "ZUPAN", "ZUPAS", "ZUPPA", "ZURFS", "ZUZIM", "ZYGAL", "ZYGON", "ZYMES", "ZYMIC"];
```
Comment (dskarbek) [inline wordlespace.html:19]: Is putting it all on one-line an important change, or is this just part of your auto-formatting?

File: wordlespace.html [hunk 2]
```diff
@@ -1,318 +1,23 @@
...
-  window.dataLayer = window.dataLayer || [];
-  function gtag(){dataLayer.push(arguments);}
-  gtag('js', new Date());
+  <!-- Global site tag (gtag.js) - Google Analytics -->
+  <script async src="https://www.googletagmanager.com/gtag/js?id=G-JBF2M8YCKP"></script>
+  <script>
+    window.dataLayer = window.dataLayer || [];
+    function gtag() { dataLayer.push(arguments); }
+    gtag('js', new Date());
+
```
Comment (dskarbek) [inline wordlespace.html]: should take this blank line out.

File: wordlespace.html [hunk 3]
```diff
@@ -332,50 +37,45 @@
       let grays = getLetters($('#gray'), 'grayInput');
       let letterFreq = new Map();
       let words = wordlist
-          .map(w => w.toUpperCase())
```
Comment (dskarbek) [inline wordlespace.html:335]: do we not need to upper-case the words?

File: wordlespace.html [hunk 4]
```diff
@@ -332,50 +37,45 @@
...
-          // must contain all yellow letters somewhere
-          .filter(w => {
-            for (let i = 0; i < allYellows.length; i += 1) {
-              if (!w.includes(allYellows[i])) {
+        // must contain all yellow letters somewhere
+        .filter(w => {
+          let matchesYellows = allYellows.every(yellow => w.includes(yellow));
```
Comment (dskarbek) [inline wordlespace.html]: why assign this to a variable, why not just in-line this in the `if(!...) return false`?

File: wordlespace.html [hunk 5]
```diff
@@ -332,50 +37,45 @@
...
+          }
+          return true;
+        });
 
       // count letter frequencies and sort by count, highest to lowest
       words.forEach(w => new Set([...w]).forEach(ltr => countLetter(letterFreq, ltr)));
       letterFreq = new Map([...letterFreq].sort(([k1, v1], [k2, v2]) => v2 - v1));
 
       $('#frequency').html(Array.from(letterFreq.entries(), ([letter, count]) => '<li>' + letter + ': ' + count + '</li>').join(''));
+      $('#words').toggle(this.checked);
```
Comment (dskarbek) [inline wordlespace.html:76]: hmm, what's the advantage of doing this here?  I assume this goes hand in hand with removing `checked` from the input in the initial HTML.  Is the point that it's hidden until after it's generated the first time?

File: wordlespace.html (hunk 1 above)
Comment (Kvadratni) [reply wordlespace.html:19]: formating

File: wordlespace.html (hunk 5 above)
Comment (Kvadratni) [reply wordlespace.html:76]: yes. we forgo rendering the whole list until the first clue.

File: wordlespace.html (hunk 4 above)
Comment (Kvadratni) [reply wordlespace.html]: I expected to use it later in the filter, but yeah no need

File: wordlespace.html (hunk 3 above)
Comment (Kvadratni) [reply wordlespace.html:335]: I've upper cased them in the array.

Comment (dskarbek) [reply wordlespace.html:335]: oh, got it.

**Comment by dskarbek** [comment]
Thanks for the input / ideas.  Applied pieces of this PR in separate commits.
https://github.com/dskarbek/wordlespace/commit/e3905daedc57811862124d91387876ed62210466
https://github.com/dskarbek/wordlespace/commit/4dc7168ce342a01106a05f596e8a20905571fda7

**Pull request closed without merging**